### DISCUSSION TITLE
Unable to install khoj backend : pip and correct python version install

### DISCUSSION THREAD

**Discussion by parimalpuranik** [khoj-ai/khoj#209] [Q&A] [answered]
PS C:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages> py --version
Python 3.10.0
PS C:\Users\xxxxxxxxxx> py -m ensurepip --upgrade
Looking in links: c:\Users\xxxxxxxxxx\AppData\Local\Temp\tmplppbbsgx
Requirement already satisfied: setuptools in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (57.4.0)
Requirement already satisfied: pip in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (21.2.3)
PS C:\Users\xxxxxxxxxx> cd c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages
PS C:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages> pip install khoj-assistant
Fatal error in launcher: Unable to create process using '"C:\Users\xxxxxxxxxx\AppData\Local\Programs\Python\Python311\python.exe"  "C:\Users\xxxxxxxxxx\AppData\Local\Programs\Python\Python311\Scripts\pip.exe" install khoj-assistant': The system cannot find the file specified.

**Comment by parimalpuranik** [comment]
I noticed that the pip was in a directory with version 3.1.1 so I deleted that directory and checked again in the 3.10 directory. Now even pip cannot be found : 

PS C:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages> py --version
Python 3.10.0
PS C:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages> py -m ensurepip --upgrade
Looking in links: c:\Users\xxxxxxxxxx\AppData\Local\Temp\tmpo2hs4ex4
Requirement already satisfied: setuptools in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (57.4.0)
Requirement already satisfied: pip in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (21.2.3)
PS C:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages> pip install khoj-assistant
pip : The term 'pip' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:1
+ pip install khoj-assistant
+ ~~~
    + CategoryInfo          : ObjectNotFound: (pip:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException

**Reply by debanjum** [reply]
> Now even pip cannot be found :

Does running `py -m ensurepip --upgrade` get back pip for you? You may also (re-)open terminal to find pip. 

Also ensure you're trying to install khoj using pip associated with python version 3.10 or lower. Use `py -m pip install khoj-assistant` to install khoj using pip associated with your default python (which seems to be python 3.10)

**Reply by parimalpuranik** [reply]
Yes, in the first comment I confirmed that pip was using 3.10 version of python. I reopened in a new window and retried with py -m pip install khoj-assistant when it worked.  I think your documentation at https://github.com/debanjum/khoj#1-Install could be updated to ask to use "py -m pip install khoj-assistant" instead of  "pip install khoj-assistant " which it currently says.  still having trouble starting khoj will keep tinkering.

**Reply by debanjum** [reply]
> I think your documentation at https://github.com/debanjum/khoj#1-Install could be updated to ask to use "py -m pip install khoj-assistant" instead of "pip install khoj-assistant " which it currently says.

Yeah, sounds like a good idea for sure. I'll do that in a bit!

> still having trouble starting khoj will keep tinkering.

Can you share what issue you're facing starting khoj? Maybe I (or others) can help you get started with using khoj faster

**Reply by parimalpuranik** [reply]
When I run py -m pip install khoj-assistant && khoj --no-gui , it first confirm khoj components are instantlled but then fails with  : 

'khoj' is not recognized as an internal or external command, operable program or batch file. 

Where does the khoj binary get installed ? 

C:\Users\xxxxxxxxxx>py -m pip install khoj-assistant && khoj --no-gui
Requirement already satisfied: khoj-assistant in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (0.6.0)
Requirement already satisfied: jinja2==3.1.2 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from khoj-assistant) (3.1.2)
Requirement already satisfied: tiktoken>=0.3.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from khoj-assistant) (0.3.3)
Requirement already satisfied: aiohttp==3.8.4 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from khoj-assistant) (3.8.4)
Requirement already satisfied: openai>=0.27.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from khoj-assistant) (0.27.4)
Requirement already satisfied: pillow==9.3.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from khoj-assistant) (9.3.0)
Requirement already satisfied: dateparser==1.1.1 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from khoj-assistant) (1.1.1)
Requirement already satisfied: uvicorn==0.17.6 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from khoj-assistant) (0.17.6)
Requirement already satisfied: torch==1.13.1 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from khoj-assistant) (1.13.1)
Requirement already satisfied: pyqt6==6.3.1 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from khoj-assistant) (6.3.1)
Requirement already satisfied: rich>=13.3.1 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from khoj-assistant) (13.3.3)
Requirement already satisfied: defusedxml==0.7.1 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from khoj-assistant) (0.7.1)
Requirement already satisfied: pyyaml==6.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from khoj-assistant) (6.0)
Requirement already satisfied: pydantic==1.9.1 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from khoj-assistant) (1.9.1)
Requirement already satisfied: tenacity>=8.2.2 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from khoj-assistant) (8.2.2)
Requirement already satisfied: fastapi==0.77.1 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from khoj-assistant) (0.77.1)
Requirement already satisfied: sentence-transformers==2.2.2 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from khoj-assistant) (2.2.2)
Requirement already satisfied: schedule==1.1.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from khoj-assistant) (1.1.0)
Requirement already satisfied: multidict<7.0,>=4.5 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from aiohttp==3.8.4->khoj-assistant) (6.0.4)
Requirement already satisfied: aiosignal>=1.1.2 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from aiohttp==3.8.4->khoj-assistant) (1.3.1)
Requirement already satisfied: attrs>=17.3.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from aiohttp==3.8.4->khoj-assistant) (22.2.0)
Requirement already satisfied: async-timeout<5.0,>=4.0.0a3 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from aiohttp==3.8.4->khoj-assistant) (4.0.2)
Requirement already satisfied: yarl<2.0,>=1.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from aiohttp==3.8.4->khoj-assistant) (1.8.2)
Requirement already satisfied: frozenlist>=1.1.1 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from aiohttp==3.8.4->khoj-assistant) (1.3.3)
Requirement already satisfied: charset-normalizer<4.0,>=2.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from aiohttp==3.8.4->khoj-assistant) (3.1.0)
Requirement already satisfied: pytz in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from dateparser==1.1.1->khoj-assistant) (2023.3)
Requirement already satisfied: python-dateutil in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from dateparser==1.1.1->khoj-assistant) (2.8.2)
Requirement already satisfied: tzlocal in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from dateparser==1.1.1->khoj-assistant) (4.3)
Requirement already satisfied: regex!=2019.02.19,!=2021.8.27,<2022.3.15 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from dateparser==1.1.1->khoj-assistant) (2022.3.2)
Requirement already satisfied: starlette==0.19.1 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from fastapi==0.77.1->khoj-assistant) (0.19.1)
Requirement already satisfied: MarkupSafe>=2.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from jinja2==3.1.2->khoj-assistant) (2.1.2)
Requirement already satisfied: typing-extensions>=3.7.4.3 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from pydantic==1.9.1->khoj-assistant) (4.5.0)
Requirement already satisfied: PyQt6-sip<14,>=13.4 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from pyqt6==6.3.1->khoj-assistant) (13.4.1)
Requirement already satisfied: PyQt6-Qt6>=6.3.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from pyqt6==6.3.1->khoj-assistant) (6.4.3)
Requirement already satisfied: transformers<5.0.0,>=4.6.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from sentence-transformers==2.2.2->khoj-assistant) (4.27.4)
Requirement already satisfied: tqdm in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from sentence-transformers==2.2.2->khoj-assistant) (4.65.0)
Requirement already satisfied: torchvision in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from sentence-transformers==2.2.2->khoj-assistant) (0.14.1)
Requirement already satisfied: numpy in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from sentence-transformers==2.2.2->khoj-assistant) (1.24.2)
Requirement already satisfied: scikit-learn in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from sentence-transformers==2.2.2->khoj-assistant) (1.2.2)
Requirement already satisfied: scipy in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from sentence-transformers==2.2.2->khoj-assistant) (1.10.1)
Requirement already satisfied: nltk in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from sentence-transformers==2.2.2->khoj-assistant) (3.8.1)
Requirement already satisfied: sentencepiece in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from sentence-transformers==2.2.2->khoj-assistant) (0.1.97)
Requirement already satisfied: huggingface-hub>=0.4.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from sentence-transformers==2.2.2->khoj-assistant) (0.13.4)
Requirement already satisfied: anyio<5,>=3.4.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from starlette==0.19.1->fastapi==0.77.1->khoj-assistant) (3.6.2)
Requirement already satisfied: h11>=0.8 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from uvicorn==0.17.6->khoj-assistant) (0.14.0)
Requirement already satisfied: click>=7.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from uvicorn==0.17.6->khoj-assistant) (8.1.3)
Requirement already satisfied: asgiref>=3.4.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from uvicorn==0.17.6->khoj-assistant) (3.6.0)
Requirement already satisfied: sniffio>=1.1 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from anyio<5,>=3.4.0->starlette==0.19.1->fastapi==0.77.1->khoj-assistant) (1.3.0)
Requirement already satisfied: idna>=2.8 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from anyio<5,>=3.4.0->starlette==0.19.1->fastapi==0.77.1->khoj-assistant) (3.4)
Requirement already satisfied: colorama in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from click>=7.0->uvicorn==0.17.6->khoj-assistant) (0.4.6)
Requirement already satisfied: packaging>=20.9 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from huggingface-hub>=0.4.0->sentence-transformers==2.2.2->khoj-assistant) (23.0)
Requirement already satisfied: filelock in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from huggingface-hub>=0.4.0->sentence-transformers==2.2.2->khoj-assistant) (3.11.0)
Requirement already satisfied: requests in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from huggingface-hub>=0.4.0->sentence-transformers==2.2.2->khoj-assistant) (2.28.2)
Requirement already satisfied: urllib3<1.27,>=1.21.1 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from requests->huggingface-hub>=0.4.0->sentence-transformers==2.2.2->khoj-assistant) (1.26.15)
Requirement already satisfied: certifi>=2017.4.17 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from requests->huggingface-hub>=0.4.0->sentence-transformers==2.2.2->khoj-assistant) (2022.12.7)
Requirement already satisfied: pygments<3.0.0,>=2.13.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from rich>=13.3.1->khoj-assistant) (2.14.0)
Requirement already satisfied: markdown-it-py<3.0.0,>=2.2.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from rich>=13.3.1->khoj-assistant) (2.2.0)
Requirement already satisfied: mdurl~=0.1 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from markdown-it-py<3.0.0,>=2.2.0->rich>=13.3.1->khoj-assistant) (0.1.2)
Requirement already satisfied: tokenizers!=0.11.3,<0.14,>=0.11.1 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from transformers<5.0.0,>=4.6.0->sentence-transformers==2.2.2->khoj-assistant) (0.13.3)
Requirement already satisfied: joblib in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from nltk->sentence-transformers==2.2.2->khoj-assistant) (1.2.0)
Requirement already satisfied: six>=1.5 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from python-dateutil->dateparser==1.1.1->khoj-assistant) (1.16.0)
Requirement already satisfied: threadpoolctl>=2.0.0 in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from scikit-learn->sentence-transformers==2.2.2->khoj-assistant) (3.1.0)
Requirement already satisfied: tzdata in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from tzlocal->dateparser==1.1.1->khoj-assistant) (2023.3)
Requirement already satisfied: pytz-deprecation-shim in c:\users\xxxxxxxxxx\appdata\local\programs\python\python310\lib\site-packages (from tzlocal->dateparser==1.1.1->khoj-assistant) (0.1.0.post0)
WARNING: You are using pip version 21.2.3; however, version 23.0.1 is available.
You should consider upgrading via the 'C:\Users\xxxxxxxxxx\AppData\Local\Programs\Python\Python310\python.exe -m pip install --upgrade pip' command.
'khoj' is not recognized as an internal or external command,
operable program or batch file.

C:\Users\xxxxxxxxxx>

**Reply by debanjum** [reply, marked as answer]
> 'khoj' is not recognized as an internal or external command,
operable program or batch file.

Ah, I see. The path at which khoj is installed isn't searched by default on your Windows. 
So you'd need to either add the default path python pip packages are installed to your PATH or call `khoj` using the whole path to it. See [my reply to this discussion](https://github.com/debanjum/khoj/discussions/194#discussioncomment-5445910) for solution.

Let me know if that doesn't work for you

**Comment by parimalpuranik** [comment]
Found the khoj.exe in scripts folder and I was able to get to configure screen.  Thanks for your help !
